### PR TITLE
hints: Reduce hint sending parallelism

### LINKED ISSUE #4974: Hinted handoff: take the number of nodes and shards in the cluster into an account when determining the maximum send parallelism
HEAD: e61a86bbb25f9e0943387738f1e138743c8fe166

## Problem
Today we set the maximum parallelism to 128 for each shard. 
This is already too many since for a medium sized node with 30 shards this gives us a 3840 in-flight writes from each node.

If on top of that all nodes send to the same destination (e.g. after it booted up after being down for some time) and taking into an account that this all sends may land on the same shard at the destination (since we still don't have a shard-awareness on a storage proxy level) this may lead to a very unpleasant situation.

## Proposal
Take into an account that with HH the situation when hints are sent to a particular node (e.g. because it was temprorarily slow or DOWN) is quite a common case.

If so, then even if there is only a single node that sends to another node, the parallelism of 128/shard is already too many.

10-20 per-destination shard would be a much more reasonable value.

But we should also take into an account the case when there are more than a single node sending hints to a single node.

So, we should divide that 10-20 hints per shard by `N-1` where N is the number of hintable nodes.

### PR DESCRIPTION
**Pull request by piodul**
Currently, we impose a limit of 128 hints to be sent simultaneously by one shard. This patch changes this limit in the following ways:
- There is a separate limit for each pair (source shard, destination node),
- Each pair is limited to `20 / (number of known live nodes - 1)` hints sent simultaneously.

Fixes #4974.

Tests:
- unit(release)
- dtest: hintedhandoff_additional_test(release)

### CODE REVIEW

File: db/hints/resource_manager.cc [hunk 1]
```diff
@@ -59,8 +59,13 @@ future<semaphore_units<semaphore_default_exception_factory>> resource_manager::g
     }
     seastar::semaphore& limiter = *it->second;
 
+    // Take into account that hints may be sent in parallel by many nodes. In the worst case all other nodes may be
+    // sending hints to the same shard on the same node, so scale by live members count (excluding the destination).
+    const size_t live_members_count = _gossiper_ptr->get_live_members_count();
+    const size_t scale_factor = (live_members_count > 1) ? live_members_count - 1 : 1;
```
Comment (denesb) [inline db/hints/resource_manager.cc]: `const size_t scale_factor = std::max(live_members_count - 1, 1);`

Would make the intent more apparent.

Comment (piodul) [reply db/hints/resource_manager.cc]: I wrote it like this in order to avoid underflow in rare case when `live_members_count == 0` (`size_t` is unsigned). I think it could happen when all other nodes are not alive, and we are shutting down. I can add a comment to make it more clear.

Another option would be a compromise: `const size_t scale_factor = std::max(live_members_count, 2) - 1;`...

...or not to care about underflow at all (nothing should break, we will just limit sending to 1 hint per endpoint), but I'd rather avoid over/underflows if possible.

File: gms/gossiper.cc [hunk 2]
```diff
@@ -797,6 +797,18 @@ std::set<inet_address> gossiper::get_live_members() {
     return live_members;
 }
 
+size_t gossiper::get_live_members_count() const {
```
Comment (psarna) [inline gms/gossiper.cc]: git grep says that there's one more usage of `get_live_members().size()` idiom in `service/storage_service.cc`, perhaps it should be migrated to `get_live_members_count()` as well.

File: db/hints/resource_manager.cc (hunk 1 above)
Comment (denesb) [reply db/hints/resource_manager.cc]: I think we should avoid under/over flow, so either the current code or the `std::max()` version is fine.

Comment (piodul) [reply db/hints/resource_manager.cc]: OK, I added a comment.

File: gms/gossiper.cc (hunk 2 above)
Comment (piodul) [reply gms/gossiper.cc]: Good catch, not sure why I didn't find it before. Changed.

**Comment by vladzcloudius** [comment]
@piodul While this PR implements what I wrote in the opening message of #4974 and I think it's probably a right thing to do as a short term solution it doesn't address the many-to-one issue I've described in the "Remarks" right after that.

It would also be nice to see some test results demonstrating positive and negative aspects of this change. In particular, check if this is going to cause the MV backlog to grow significantly due to HH not being able to send out as quickly as before. 
Or see if these patches are going to improve the situation when the node that was DOWN gets UP and all start sending hints to it (a little more about that below).

Now to what this PR *doesn't do* and what I think is probably more important.

Even if you limit the maximum outstanding hints sent from a single shard by 1 you can still have an issue because the problem is not only in the amount of hints being sent. In the many-to-one case the main issue is the amount of hints being sent to the same node from all other nodes and no matter how you limit a single sender it would still be linear to the number of senders.

IMO the much better approach would be to implement what I've described in the presentation, which is:

1) Make HH use a separate TCP socket.
2) Make HH use a separate RPC verb (probably you may use the same verb if it arrives on a separate connection) and make sure it's handled in the "streaming" priority class.

This will already make sure that initial sending parallelism is not going to impact the users requests and will also address the many-to-one problem because a throttling on each sender is going to be applied by the receiver by means of TCP window.

The only thing that would still be an issue is memory allocation on the sender side. - Again, even if you limit the maximum outstanding sent hint by 1 per-shard you can still don't resolve the memory problem in a general case because one huge mutation may require more memory than 128 small mutations. 

We need a better approach here - the straight forward one would be to use memory pools (could be slabs). Use a fixed size pool for HH for instance.

A more complicated but probably more promising approach would be to use some memory controllers that would do similar things to what our LSA-based components (cache/memtables) do today - takes memory from a lower tier agents to higher tier agent when needed and giving (almost) all memory to a lower tier agents when a higher tier doesn't need it in order to let them go fast.

**Comment by vladzcloudius** [comment]
I refreshed my memory and I noticed that we've already handled the senders memory allocation. So, this is not even an issue. 
So, back to my point - this whole patch may be dropped if you implement two points I mentioned above.

**Comment by piodul** [comment]
@vladzcloudius Thanks for your comments. I agree that the solution with separate TCP socket sounds much better. I will open a new pull request when I implement it.

File: db/hints/manager.cc [hunk 3]
```diff
@@ -684,7 +684,7 @@ future<> manager::end_point_hints_manager::sender::send_one_mutation(frozen_muta
 }
 
 future<> manager::end_point_hints_manager::sender::send_one_hint(lw_shared_ptr<send_one_file_ctx> ctx_ptr, fragmented_temporary_buffer buf, db::replay_position rp, gc_clock::duration secs_since_file_mod, const sstring& fname) {
-    return _resource_manager.get_send_units_for(buf.size_bytes()).then([this, secs_since_file_mod, &fname, buf = std::move(buf), rp, ctx_ptr] (auto units) mutable {
+    return _resource_manager.get_send_units_for(_ep_key, buf.size_bytes()).then([this, secs_since_file_mod, &fname, buf = std::move(buf), rp, ctx_ptr] (auto units) mutable {
         // Future is waited on indirectly in `send_one_file()` (via `ctx_ptr->file_send_gate`).
```
Comment (avikivity) [inline db/hints/manager.cc]: Doesn't this make memory used by hints scale with the cluster size?

Let's say a table uses 5MB rows (large but not crazy). Then a 100-node cluster will require 500MB per shard just for in-flight hints.

Comment (avikivity) [reply db/hints/manager.cc]: And btw the commit description should explain why you make the change, not just what the change is.

File: db/hints/resource_manager.cc [hunk 4]
```diff
@@ -59,8 +59,15 @@ future<semaphore_units<semaphore_default_exception_factory>> resource_manager::g
     }
     seastar::semaphore& limiter = *it->second;
```
Comment (avikivity) [inline db/hints/resource_manager.cc]: What's the problem with "all but one nodes simultaneously send hints to the remaining node"?

Note that 20 may be too little for small hints (bandwidth will be limited by round-trip latency) and too much for large hints (20*5MB = 100MB). Ultimately we should use rpc streams for such cases, streams eliminate the dependency on round-trip latency and place bounds on memory usage. @gleb-cloudius @asias

**Comment by avikivity** [comment]
Cover letter should explain the motivation. I guess it is about reducing memory usage for in-flight hints, but I think it is doing the opposite for large clusters.

**Comment by piodul** [comment]
@vladzcloudius I have opened a new PR here: https://github.com/scylladb/scylla/pull/5378

I'm closing this PR right now.
